wandb/apis/public.py

```diff
@@ -376,6 +380,9 @@ def api_key(self):
         # Environment should take precedence
         if os.getenv("WANDB_API_KEY"):
             key = os.environ["WANDB_API_KEY"]
+        # Constructor value should take further precedence
+        if hasattr(self, "_api_key"):
+            key = self._api_key
```

Comment:
I think it would actually be better to make this method memoized.  It would definitely be super confusing if during a jupyter session or python process the key changed, i.e.

```python
def api_key(self):
    # You'll need to be sure self._api_key is atleast set to None in the __init__
    if self._api_key is not None:
        return self._api_key
    auth = requests.utils.get_netrc_auth(self.settings["base_url"])
    key = None
    if auth:
        key = auth[-1]
    # Environment should take precedence
    if os.getenv("WANDB_API_KEY"):
        key = os.environ["WANDB_API_KEY"]
    self._api_key = key
    return self._api_key
```

Reply:
sounds good 👍

Reply:
done in [fd1dce4](https://github.com/wandb/client/pull/3657/commits/fd1dce4c3676ab59e7cf79cb3658ad5679ddc6aa)